compiler/sem/sem.nim

```diff
@@ -167,7 +167,9 @@ proc deltaTrace(stopProc, indent: string, entries: seq[StackTraceEntry])
 template semIdeForTemplateOrGenericCheck(conf, n, requiresCheck) =
   # we check quickly if the node is where the cursor is
   when defined(nimsuggest):
-    if n.info.fileIndex == conf.m.trackPos.fileIndex and n.info.line == conf.m.trackPos.line:
+    if conf.m.trackPos.col != -1 and
```

Comment:
```suggestion
    if conf.m.trackPosAttached and
```

Reply:
while looking at the `lexer.nim`  `L.config.ideCmd in {ideSug, ideCon}`, does this mean `semIdeForTemplateOrGeneric` only used for these commands?

Reply:
D'oh, you're right, I completely misunderstood the meaning of `trackPosAttached`, sorry.

Hm, I think the original intention with `semIdeForTemplateOrGeneric` was to allow *suggestions* within templates/generics (basic highlighting, at least of the unresolved symbols, is already taken care of through `onMarkUsed` -> `suggestSym`), so making that explicit, via:

```suggestion
    if conf.ideCmd in {ideSug, ideCon} and
```

should be fine here. @saem: What are your thoughts on this?

Reply:
I still think with col validation is better, as their line value are default to 0, we don't know it's default value or real first line. I'd like use these two conditions all

Reply:
That's a very indirect test. `semIdeForTemplateOrGeneric` doesn't care about the track position's column value nor should it -- it's interested in whether the cursor has meaning.

Testing for 'A' directly is better than testing for 'B', where 'B' is *currently* implied by 'A'. Both may yield the same result at the time, but testing for 'B' introduces the hidden expectation that 'A' continues to imply 'B'.

Reply:
well,  I'll think it just test `trackPos` is valid

Reply:
Could you elaborate? I don't understand what you mean.

Reply:
Or, asked differently, why do you think that the `trackPos.col != -1` test is better than `ideCmd` in {ideSug, ideCon}`?

Reply:
the list of commands where we need a cursor might be a bigger list: usage, definition, definition + usage, suggest, consider_parameter. I believe we _must_ have a cursor for all of those, so if the cursor isn't valid we should not act in the first place. Assuming that's correct, I think cursor validation (`trackPos` etc) should happen early, if we don't have it for the relevant command then we don't even bother executing the command. After that then `semIdeForTemplateOrGeneric` should test based on the command.

Thoughts from both of you about the above?

Reply:
> Or, asked differently, why do you think that the `trackPos.col != -1` test is better than `ideCmd in {ideSug, ideCon}`?

I'll use them all

Reply:
> I'll use them all, it's not about which on e is better, `trackPos.col != -1` is just check `trackPos` is valid.

Like @saem said, if `trackPos` is not valid, the command shouldn't be executed in the first place.

Reply:
okay